### ISSUE TITLE
GitVersion altering files

### ISSUE THREAD

**Issue by James-DBA-Anderson** [appveyor/ci#2976]
Hi,

I'm using GitVersion and it seems to be behaving strangely with GitHub PR builds.

The AppVeryor build is doing:

```
git clone -q git@github.com:X/X.git C:\Projects\X
git fetch -q origin +refs/pull/529/merge:
git checkout -qf FETCH_HEAD
```
as normal.

I then have a step in my AppVeyor.yml to run GitVersion. 
```
gitversion $env:APPVEYOR_BUILD_FOLDER /l console /output buildserver /nofetch /b $env:APPVEYOR_REPO_BRANCH 
```

I have a step after this that runs my new test-env-build script. This step works on the branch build but fails on the PR build as it is unable to find the new test-env-build script.

To see whats going on, I added a step to run ```ls``` before the Gitversion command and another ```ls``` after.

The test-env-build script is there to start with but it's not there after the GitVersion command runs.

It looks like it is switching to another branch or rolling back to a previous commit.

Any ideas on this would be really appreciated.

**Comment by gep13** [comment]
@James-DBA-Anderson what version of GitVersion are you using?

In order for GitVersion to work properly, and to assert the correct semantic version number, it needs to have all the history of the git repository.  When running on a build server, that history isn't always available, and as a result, GitVersion runs a process known as branch normalization, to pull down all the required information.  It sounds like for whatever reason, this is leaving the currently git repository in a state that isn't desired.

Can you confirm what is in the two environment variables that you are using in your execution of gitversion?

**Comment by James-DBA-Anderson** [comment]
@gep13 I'm using the lastest VS2017 AppVeyor image which I assume has v4 of gitversion installed. Can't see what it is in the release notes.

Values of env vars:
$env:APPVEYOR_BUILD_FOLDER = C:\Projects\GitVersionTest
$env:APPVEYOR_REPO_BRANCH = master

**Comment by gep13** [comment]
@James-DBA-Anderson is there a reason that you are attempting to specify a branch name in the execution of GitVersion?  I have never found this to be necessary when running on AppVeyor and the repo comes from GitHub.

**Comment by James-DBA-Anderson** [comment]
No, and I'm not sure what the /b switch does. I'm just using the command that GitVersion init gae me after going through the wizard. I'll try it without.

**Comment by IlyaFinkelshteyn** [comment]
Also I guess `$env:APPVEYOR_PULL_REQUEST_HEAD_REPO_BRANCH` may work for PR builds. Non-empty `$env:APPVEYOR_PULL_REQUEST_HEAD_REPO_BRANCH` means that build is a PR one, so you can write a conditional statement like this: 

```
$BRANCH_TO_USE = if ($env:APPVEYOR_PULL_REQUEST_HEAD_REPO_BRANCH) {$env:APPVEYOR_PULL_REQUEST_HEAD_REPO_BRANCH} else {$env:APPVEYOR_REPO_BRANCH}

gitversion $env:APPVEYOR_BUILD_FOLDER /l console /output buildserver /nofetch /b $BRANCH_TO_USE 
```

but again, as @gep13 said, this /b switch probably is not necessary.
